### DISCUSSION TITLE
Couldn't connect VM using SSH

### DISCUSSION THREAD

**Discussion by ghost** [utmapp/UTM#2535] [UTM Help (macOS)]
I have installed UTM on my Mac M1 and created a virtual machine using the RHEL7.6 ISO image.
The installation of the virtual machine was successful, but when I try to connect to the created virtual machine from my local machine using port forwarding, I get the following error and it fails.

$ ssh root@localhost
kex_exchange_identification: read: Connection has been reset by peer

I would like to know how to solve this.

[Environment]
OS: macOS Big Sur
Architecture M1
UTM Version: 2.0.28 (28)

**Comment by ghost** [comment]
try setting up port forwarding to a different port than ssh (i. e. port 4321), and set up sshd in RHEL to listen on that port, then try this
`$ ssh root@localhost -p <port>`

**Reply by TnGadaria** [reply]
Just create a normal user in guest Linux and ssh with that user (not with root). you are allowed to give sudo privilege to the user.

**Comment by ghost** [comment]
Hi
Thank you for your comment.

I have configured port 4321 on the local PC to be port forwarded to port 22 on the virtual machine.
![image](https://user-images.githubusercontent.com/84234456/118381817-e95c8080-b629-11eb-8251-555cc44ca504.png)

However, I get the same error.
$  ssh -v root@localhost -p 4321
OpenSSH_8.1p1, LibreSSL 2.7.3
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 47: Applying options for *
debug1: Connecting to localhost port 4321.
debug1: Connection established.
debug1: identity file /Users/minamii0425/.ssh/id_rsa type 0
debug1: identity file /Users/minamii0425/.ssh/id_rsa-cert type -1
debug1: identity file /Users/minamii0425/.ssh/id_dsa type -1
debug1: identity file /Users/minamii0425/.ssh/id_dsa-cert type -1
debug1: identity file /Users/minamii0425/.ssh/id_ecdsa type -1
debug1: identity file /Users/minamii0425/.ssh/id_ecdsa-cert type -1
debug1: identity file /Users/minamii0425/.ssh/id_ed25519 type -1
debug1: identity file /Users/minamii0425/.ssh/id_ed25519-cert type -1
debug1: identity file /Users/minamii0425/.ssh/id_xmss type -1
debug1: identity file /Users/minamii0425/.ssh/id_xmss-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.1
kex_exchange_identification: read: Connection reset by peer

Port 22 is being listened on the virtual machine.
![image](https://user-images.githubusercontent.com/84234456/118381710-d72e1280-b628-11eb-95ba-49f40d7dce65.png)

Where else should I check?

**Reply by ghost** [reply]
are your ssh keys setup correctly?

**Comment by ghost** [comment]
The settings around SSH are as follows.

[Mac m1]
minamii0425@MinaminoAir .ssh % ls -l
total 16
-rw-------  1 minamii0425  staff  2610  5 15 17:23 id_rsa
-rw-------  1 minamii0425  staff   577  5 15 17:23 id_rsa.pub
minamii0425@MinaminoAir .ssh %    

[VM]
![image](https://user-images.githubusercontent.com/84234456/118394697-30299500-b681-11eb-8fff-75d7f507b2ef.png)

The content of id_rsa.pub on Mac m1 is posted to ~/.ssh/authorized_keys on VM.

*Changed the port used from 4321 to 5432.
minamii0425@MinaminoAir .ssh % ssh -v minamii0425@localhost -p 5432
OpenSSH_8.1p1, LibreSSL 2.7.3
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 47: Applying options for *
debug1: Connecting to localhost port 5432.
debug1: Connection established.
debug1: identity file /Users/minamii0425/.ssh/id_rsa type 0
debug1: identity file /Users/minamii0425/.ssh/id_rsa-cert type -1
debug1: identity file /Users/minamii0425/.ssh/id_dsa type -1
debug1: identity file /Users/minamii0425/.ssh/id_dsa-cert type -1
debug1: identity file /Users/minamii0425/.ssh/id_ecdsa type -1
debug1: identity file /Users/minamii0425/.ssh/id_ecdsa-cert type -1
debug1: identity file /Users/minamii0425/.ssh/id_ed25519 type -1
debug1: identity file /Users/minamii0425/.ssh/id_ed25519-cert type -1
debug1: identity file /Users/minamii0425/.ssh/id_xmss type -1
debug1: identity file /Users/minamii0425/.ssh/id_xmss-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.1
kex_exchange_identification: read: Connection reset by peer

**Comment by conath** [comment]
The solution for Mac users is to change the network mode to shared or bridged mode. The emulated network seems to break SSH.

**Reply by itsgodzilla** [reply]
when emulated network(for installing x86 ubuntu), just add a vlan network device, and remove the old one. Then edit portforward-"TCP->22->1022". That works fine for me to ssh the x86 vm from local Mac M1.
The key is just hold one network device on vm.

**Comment by fffx** [comment]
I'm using UTM on a Macbook m1. Installed two ubuntu servers(20.4), with x86 and arm respectively,   both with emulated network, and mapped 22(guest) to 2222(host).

The x86's port map of ssh works out of box, but the arm version doesn't work. I finally found out that my interface(enp0s7) is not properly configured automatically on start. **There is no IP address assigned to this interface**. So I updated my network config like this

```
# /etc/systemd/network/20-wired.network
[Match]
Name=enp0s7

[Network]
DHCP=yes
```
reboot,  then it works
```
$ ip addr

1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: enp0s7: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether de:d8:2c:00:ca:3b brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global dynamic enp0s7
       valid_lft 85143sec preferred_lft 85143sec
    inet6 fec0::dcd8:2cff:fe00:ca3b/64 scope site dynamic mngtmpaddr
       valid_lft 86227sec preferred_lft 14227sec
    inet6 fe80::dcd8:2cff:fe00:ca3b/64 scope link
       valid_lft forever preferred_lft foreve
```

My network config may not be the best practice, you better follow your according distribution documentation

**Reply by brannerchinese** [reply]
I'm surprised to see the original poster's screenshot, as the network settings in UTM (v. 3.0.4 for MacOS) look very different in my installation:

![utm-vm-network-settings-3 0 4-(46)](https://user-images.githubusercontent.com/549884/156983138-0e352905-a5cf-41f1-8601-d0f2e281a7a2.jpg)

I don't see an option for port-forwarding at all. Am I missing something?

**Reply by osy** [reply]
You're replying to a post from 2021 when the version is 2.0. It's better to make a new post as most things here are probably outdated.

To answer your question though, you don't have port forwarding because you chose shared network not emulated vlan. Port forwarding isn't supported/doesn't make sense for shared network because there is no separate VLAN.

**Comment by joeljgeorge** [comment]
To confirm, is ssh still broken on emulated VLAN?

I'm able to ssh into my virtual box using this common command: ssh -p 22022 ubuntu@localhost

But then I set the Network settings to 'Emulated VLAN', and set the IP address of the Guest Network to 192.168.56.1 and DHCP start to 192.168.56.101.

Then, when I try to connect with this command: ssh -p 22 ubuntu@192.168.56.101

The command hangs and the operation eventually times out. Is this still because ssh is broken on emulated VLAN?

**Reply by conath** [reply]
No, this is not the same issue. You seem to be misunderstanding how Emulated VLAN mode works. It creates a software only subnet for your VM, which means it is inaccessible from the host or other VMs. To SSH into it you need to use UTM’s port forwarding feature to set up the port for SSH and then connect to that port on localhost.

**Reply by joeljgeorge** [reply]
Ah I see - great, thanks so much for the clear explanation!

**Comment by 7lawa9111** [comment]
I have the same issue, was struggling with it for days, appreciate if anyone was able to fix.

**Reply by Cay-Zhang** [reply]
I was able to connect after installing the ssh package and starting ssh service on my Ubuntu VM set up with emulated VLAN and port forwarding.

**Comment by I-can-do-fourier** [comment]
I have the same problem. My solution is changing the UTM network config to the shared network mode. Next, follow the method of @fffx and reboot vm.Then, check your IP in your virtual environment. Finally, use ssh to connect to this IP.

````
sudo vim /etc/systemd/network/20-wired.network
[Match]
Name=enp0s1

[Network]
DHCP=yes
````

`ip -4 addr`
<img width="757" alt="image" src="https://user-images.githubusercontent.com/55371663/208287261-ec1dbf81-726b-49b0-9e69-72d8eaddcdbb.png">

For me, I just use the following command in my host terminal:
`ssh My@192.168.64.2`

Actually, I don't know why this work and would like to know the reason.If somebody know it,Please tell me.Thanks.

**Comment by saurabh-hirani** [comment]
For those following these instructions and if it is still not working, check if you are connected to a vpn which is having a similar subnet to the guest.

**Reply by xyzst** [reply]
This was my issue using tailscale and connected to `exit node`. Disconnecting from exit node (but still connected to tailnet) allowed me to ssh into VM

**Comment by jajuTT** [comment]
Also check per‑app NetworkExtension content filters on macOS (especially on managed devices). In my case, outbound traffic from iTerm2 to the VM’s shared subnet (192.168.64.0/24 on bridge100) was blocked, so SSH failed with: `ssh: connect to host 192.168.64.x port 22: No route to host`. The same SSH from Apple’s Terminal app worked. A tcpdump on bridge100 showed no packets from iTerm2, but a normal handshake from Terminal.

**Reply by fferino-fungible** [reply]
This article and its links might be useful: https://apple.stackexchange.com/questions/476030/no-route-to-host-on-iterm2-only
